### PR TITLE
[ENG-5462] Implement Amazon S3

### PR DESCRIPTION
**Pull request by adlius**

### CODE REVIEW

File: addon_service/authorized_storage_account/serializers.py [hunk 1]
```diff
@@ -95,6 +95,13 @@ def create(self, validated_data):
             authorized_account.initiate_oauth2_flow(
                 validated_data.get("authorized_scopes")
             )
+        if (
+            external_service.credentials_format
+            is CredentialsFormats.ACCESS_KEY_SECRET_KEY
+        ):
+            authorized_account.validate_access_key_secret_key(
+                validated_data["credentials"]
+            )
```
Comment (adlius) [inline addon_service/authorized_storage_account/serializers.py:104]: Make it more generic

File: addon_imps/storage/s3.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    @classmethod
+    def validate_credentials(cls, credentials):
+        access_key = credentials.access_key
+        secret_key = credentials.secret_key
+        s3 = boto3.client(
+            "s3", aws_access_key_id=access_key, aws_secret_access_key=secret_key
+        )
+        try:
+            s3.list_buckets()
+        except exceptions.ClientError:
```
Comment (adlius) [inline addon_imps/storage/s3.py:37]: Import as `BotoExceptions`

File: addon_imps/storage/s3.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+class S3ClientImp(storage.ClientRequestorImp):
+    client: Boto3Client
+
+
+class S3StorageImp(storage.StorageAddonImp, S3ClientImp):
+    """storage on Amazon S3"""
+
+    @classmethod
+    def validate_credentials(cls, credentials):
```
Comment (adlius) [inline addon_imps/storage/s3.py:29]: `confirm_credentials`?

File: addon_toolkit/interfaces/storage.py [hunk 4]
```diff
@@ -150,4 +150,20 @@ class StorageAddonImp(AddonImp):
...
+    """base class for storage addon implementations using GV network"""
+
     network: HttpRequestor
+
+
+@dataclasses.dataclass()
+class ClientRequestorImp:
+    """base class for storage addon with custom clients"""
+
+    client: typing.Any
```
Comment (adlius) [inline addon_toolkit/interfaces/storage.py:166]:
```suggestion
    client: typing.Any = dataclasses.field(init=False)
```

File: addon_service/addon_imp/instantiation.py [hunk 5]
```diff
@@ -15,14 +17,21 @@ async def get_storage_addon_instance(
...
+            network=GravyvaletHttpRequestor(
+                client_session=await get_singleton_client_session(),
+                prefix_url=config.external_api_url,
+                account=account,
+            ),
+        )
+    else:
+        imp = imp_cls(config=config)
+    if isinstance(imp, ClientRequestorImp):
+        await imp.construct_client(account)
```
Comment (adlius) [inline addon_service/addon_imp/instantiation.py:32]:
```suggestion
    if issubclass(imp_cls, HttpRequestorImp):
        imp = imp_cls(
            config=config,
            network=GravyvaletHttpRequestor(
                client_session=await get_singleton_client_session(),
                prefix_url=config.external_api_url,
                account=account,
            ),
        )
    if isinstance(imp, ClientRequestorImp):
        imp = imp_cls(credentials=account.credentials, config=config)
```

File: addon_toolkit/interfaces/storage.py [hunk 6]
```diff
@@ -150,4 +150,20 @@ class StorageAddonImp(AddonImp):
...
+
+    def construct_client(self, account):
+        pass
```
Comment (adlius) [inline addon_toolkit/interfaces/storage.py:169]: def __post_init__(self, credentials):
    self.client = self.create_client(credentials)

File: addon_toolkit/interfaces/storage.py [hunk 7]
```diff
@@ -150,4 +150,20 @@ class StorageAddonImp(AddonImp):
...
+class HttpRequestorImp:
+    """base class for storage addon implementations using GV network"""
+
     network: HttpRequestor
+
+
+@dataclasses.dataclass()
+class ClientRequestorImp:
+    """base class for storage addon with custom clients"""
+
```
Comment (adlius) [inline addon_toolkit/interfaces/storage.py:165]: credentials: initVar[crendentials]

File: addon_toolkit/interfaces/storage.py [hunk 8]
```diff
@@ -150,4 +150,20 @@ class StorageAddonImp(AddonImp):
...
+
+@dataclasses.dataclass(frozen=True)
+class HttpRequestorImp:
+    """base class for storage addon implementations using GV network"""
+
     network: HttpRequestor
+
+
+@dataclasses.dataclass()
```
Comment (adlius) [inline addon_toolkit/interfaces/storage.py:162]:
```suggestion
@dataclasses.dataclass(frozen=True)
```

File: addon_service/authorized_storage_account/serializers.py [hunk 9]
```diff
@@ -105,7 +105,9 @@ def create(self, validated_data):
             )
         else:
             authorized_account.credentials = validated_data["credentials"]
-
+            authorized_account.imp_cls.validate_credentials(
+                authorized_account.credentials
+            )
```
Comment (adlius) [inline addon_service/authorized_storage_account/serializers.py:110]: Move this up a line.

File: addon_imps/storage/s3.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    def construct_client(self, account):
+        access_key = account._credentials.decrypted_credentials.access_key
+        secret_key = account._credentials.decrypted_credentials.secret_key
+        self.client = boto3.client(
+            "s3", aws_access_key_id=access_key, aws_secret_access_key=secret_key
+        )
+
+    async def get_item_info(self, item_id: str) -> storage.ItemResult:
+        bucket = item_id.split("/", 1)[0]
+        key = item_id.split("/", 1)[1]
```
Comment (jwalz) [inline addon_imps/storage/s3.py:50]:
```suggestion
        bucket, key = item_id.split("/", 1)
```
Does this work?

File: addon_imps/storage/s3.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                    item_type=storage.ItemType.FILE,
+                )
+            else:
+                # That means this is a folder
+                return storage.ItemResult(
+                    item_id=item_id,
+                    item_name=item_id,
+                    item_type=storage.ItemType.FOLDER,
+                )
+        return None
```
Comment (adlius) [inline addon_imps/storage/s3.py:67]: Raise an exception when the item doesn't exist?

File: addon_imps/storage/s3.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        if response["Contents"]:
+            for file in response["Contents"]:
+                results.append(
+                    storage.ItemResult(
+                        item_id=file["Key"],
+                        item_name=bucket["Key"],
+                        item_type=storage.ItemType.FILE,
+                    )
+                )
+        return
```
Comment (adlius) [inline addon_imps/storage/s3.py:104]: If we can distinguish between an id that doesn't exist and an id that has no child items, then we should raise error when the id doesn't exist, and we should return an empty list when the ide has no child items.

File: addon_imps/storage/s3.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        self,
+        item_id: str,
+        page_cursor: str = "",
+        item_type: storage.ItemType | None = None,
+    ) -> storage.ItemSampleResult:
+        bucket = item_id.split("/", 1)[0]
+        key = item_id.split("/", 1)[1]
+        response = self.client.list_objects(Bucket=bucket, Prefix=key, Delimiter="/")
+        results = []
+        if response["CommonPrefixes"]:
```
Comment (jwalz) [inline addon_imps/storage/s3.py:86]:
```suggestion
        if response["CommonPrefixes"] and item_type is not ItemType.FILE:
```

File: addon_imps/storage/s3.py [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+            for folder in response["CommonsPrefixes"]:
+                results.append(
+                    storage.ItemResult(
+                        item_id=folder["Prefix"],
+                        item_name=folder["Prefix"],
+                        item_type=storage.ItemType.FOLDER,
+                    )
+                )
+        if response["Contents"]:
```
Comment (jwalz) [inline addon_imps/storage/s3.py:95]:
```suggestion
        if response["Contents"] and item_type is not ItemType.FOLDER:
```

File: addon_imps/storage/s3.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+# from addon_service.credentials.models import ExternalCredentials
+
+
+class Boto3Client:
+    pass
+
+
+class S3ClientImp(storage.ClientRequestorImp):
+    client: Boto3Client
+
```
Comment (jwalz) [inline addon_imps/storage/s3.py:23]: My instinct is that either this should define `construct_client` or else the `client` type statement should just live in the S3StorageImp

**Review by jwalz — COMMENTED** [review]
A few more nits/suggestions on top of what came up in Open House. Thanks for the hard work here!

File: addon_toolkit/interfaces/storage.py [hunk 16]
```diff
@@ -152,18 +154,23 @@ class StorageAddonImp(AddonImp):
...
-    client: typing.Any
+    client: typing.Any = dataclasses.field(init=False)
+    credentials: dataclasses.InitVar[Credentials]
+
+    def __post_init__(self, credentials):
+        self.client = self.create_client(credentials)
 
-    def construct_client(self, account):
-        pass
+    @abstractmethod
```
Comment (jwalz) [inline addon_toolkit/interfaces/storage.py:174]: Does this decorator do anything without the the ABCMeta metaclass? My gut says we should only need either the `@abstractmethod` decorator or the `NotImplementedError`

**Review by jwalz — COMMENTED** [review]
Looking good! A couple of nits and one question

**Pull request merged**